src/storage/txn/latch.rs

```diff
@@ -5,21 +5,22 @@ use std::collections::VecDeque;
 use std::hash::{Hash, Hasher};
 use std::usize;
 
-use spin::Mutex;
+use spin::{Mutex, MutexGuard};
 
-const WAITING_LIST_SHRINK_SIZE: usize = 32;
+const WAITING_LIST_SHRINK_SIZE: usize = 8;
+const WAITING_LIST_MAX_CAPACITY: usize = 16;
 
 /// Latch which is used to serialize accesses to resources hashed to the same slot.
 ///
-/// Latches are indexed by slot IDs. The keys of a command are hashed to slot IDs, then the command
-/// is added to the waiting queues of the latches.
+/// Latches are indexed by slot IDs. The keys of a command are hashed into unsigned numbers,
+/// then the command is added to the waiting queues of the latches.
 ///
 /// If command A is ahead of command B in one latch, it must be ahead of command B in all the
 /// overlapping latches. This is an invariant ensured by the `gen_lock`, `acquire` and `release`.
 #[derive(Clone)]
 struct Latch {
-    // store waiting commands
-    pub waiting: VecDeque<u64>,
+    // store hash value of the key and command ID which required this key.
```

Comment:
required -> requires? Maybe this sentence can be written better but I don't know how either.......